src/service/viewer-impl.js

```diff
@@ -308,13 +314,34 @@ export class Viewer {
         !this.docState_.isHidden();
   }
 
+  /**
+   * Whether the AMP document has been ever visible before. Since the visiblity
+   * state of a document can be flipped back and forth we sometimes want to know
+   * if a document has ever been visible.
+   * @return {boolean}
+   */
+  hasBeenVisible() {
+    if (this.isVisible()) {
+      this.hasBeenVisible_ = true;
```

Comment:
This seems incorrect. It means that we only observe this property if we asked for it. Sounds very HMM :) In other words, if document went from H -> V -> H, and I asked after, I will never know it was visible.